### PR TITLE
Properly lower add and mul

### PR DESCRIPTION
**Pull request by wonjoolee95**
Partly fixes https://github.com/pytorch/xla/issues/6589

### CODE REVIEW

File: torch_xla/csrc/elementwise.h [hunk 1]
```diff
@@ -125,6 +125,14 @@ xla::XlaOp BuildRsub(xla::XlaOp input, xla::XlaOp other, xla::XlaOp alpha);
 // out = input − alpha * other
 xla::XlaOp BuildSub(xla::XlaOp input, xla::XlaOp other, xla::XlaOp alpha);
 
+// Compuate the add function. Adds other, scaled by alpha, from input.
+// out = input − alpha * other
+xla::XlaOp BuildAdd(xla::XlaOp input, xla::XlaOp other, xla::XlaOp alpha);
+
+// Compuate the mul function.
```
Comment (anw90) [inline torch_xla/csrc/elementwise.h:132]: typo

File: torch_xla/csrc/elementwise.h [hunk 2]
```diff
@@ -125,6 +125,14 @@ xla::XlaOp BuildRsub(xla::XlaOp input, xla::XlaOp other, xla::XlaOp alpha);
 // out = input − alpha * other
 xla::XlaOp BuildSub(xla::XlaOp input, xla::XlaOp other, xla::XlaOp alpha);
 
+// Compuate the add function. Adds other, scaled by alpha, from input.
+// out = input − alpha * other
```
Comment (anw90) [inline torch_xla/csrc/elementwise.h:129]: out = input + alpha * other

**Comment by wonjoolee95** [comment]
I noticed that one of the SPMD unit tests that checks HLOs failed previously and seems like the HLOs are slightly changed due to this PR because it updates the lowering logic. In this [simple SPMD unit test](https://github.com/pytorch/xla/blob/master/test/spmd/test_xla_sharding.py#L832), the HLO without this change is:
```
ENTRY %IrToHlo.14 (p0.4: f32[1,128], p1.5: f32[1,128]) -> (f32[1,128]) {
  %p1.5 = f32[1,128]{1,0} parameter(1), sharding={replicated}
  %p0.4 = f32[1,128]{1,0} parameter(0), sharding={replicated}
  %constant.3 = f32[] constant(1)
  %broadcast.6 = f32[1,128]{1,0} broadcast(f32[] %constant.3), dimensions={}
  %multiply.7 = f32[1,128]{1,0} multiply(f32[1,128]{1,0} %p0.4, f32[1,128]{1,0} %broadcast.6)
  %add.8 = f32[1,128]{1,0} add(f32[1,128]{1,0} %p1.5, f32[1,128]{1,0} %multiply.7)
  %custom-call.9 = f32[1,128]{1,0} custom-call(f32[1,128]{1,0} %add.8), custom_call_target="Sharding", sharding={replicated}
  %constant.2 = f32[] constant(0)
  %constant.1 = f32[] constant(1)
  %multiply.10 = f32[] multiply(f32[] %constant.2, f32[] %constant.1)
  %broadcast.11 = f32[1,128]{1,0} broadcast(f32[] %multiply.10), dimensions={}
  %add.12 = f32[1,128]{1,0} add(f32[1,128]{1,0} %custom-call.9, f32[1,128]{1,0} %broadcast.11)
  ROOT %tuple.13 = (f32[1,128]{1,0}) tuple(f32[1,128]{1,0} %add.12)
}
```
And the updated HLO looks like:
```
ENTRY %IrToHlo.14 (p0.5: f32[1,128], p1.8: f32[1,128]) -> (f32[1,128]) {
  %p1.8 = f32[1,128]{1,0} parameter(1), sharding={replicated}
  %p0.5 = f32[1,128]{1,0} parameter(0), sharding={replicated}
  %constant.4 = f32[] constant(1)
  %broadcast.6 = f32[1,128]{1,0} broadcast(f32[] %constant.4), dimensions={}
  %multiply.7 = f32[1,128]{1,0} multiply(f32[1,128]{1,0} %p0.5, f32[1,128]{1,0} %broadcast.6)
  %add.9 = f32[1,128]{1,0} add(f32[1,128]{1,0} %p1.8, f32[1,128]{1,0} %multiply.7)
  %custom-call.10 = f32[1,128]{1,0} custom-call(f32[1,128]{1,0} %add.9), custom_call_target="Sharding", sharding={replicated}
  %constant.2 = f32[] constant(0)
  %constant.1 = f32[] constant(1)
  %multiply.3 = f32[] multiply(f32[] %constant.2, f32[] %constant.1)
  %broadcast.11 = f32[1,128]{1,0} broadcast(f32[] %multiply.3), dimensions={}
  %add.12 = f32[1,128]{1,0} add(f32[1,128]{1,0} %custom-call.10, f32[1,128]{1,0} %broadcast.11)
  ROOT %tuple.13 = (f32[1,128]{1,0}) tuple(f32[1,128]{1,0} %add.12)
}
```
The contents of the HLOs are the same, the only things that differ is the suffix numbers, ex `%custom-call.9` -> `%custom-call.10`.

File: torch_xla/csrc/elementwise.cpp [hunk 3]
```diff
@@ -500,4 +500,28 @@ xla::XlaOp BuildSub(xla::XlaOp input, xla::XlaOp other, xla::XlaOp alpha) {
...
+xla::XlaOp BuildAdd(xla::XlaOp input, xla::XlaOp other, xla::XlaOp alpha) {
+  // Three-way shape and value promotion
+  std::tie(input, other) = XlaHelpers::Promote(input, other);
+  std::tie(input, alpha) = XlaHelpers::Promote(input, alpha);
+  std::tie(input, other) = XlaHelpers::Promote(input, other);
+
+  xla::XlaOp multiplied =
+      xla::Mul(other, alpha, XlaHelpers::getBroadcastDimensions(other, alpha));
+  xla::XlaOp add_result = xla::Add(
+      input, multiplied, XlaHelpers::getBroadcastDimensions(input, multiplied));
```
Comment (bhavya01) [inline torch_xla/csrc/elementwise.cpp:512]: Just curious if we need XlaHelpers::getBroadcastDimensions(input, multipled) as we promote all the tensors to be of the same shape before?

Comment (wonjoolee95) [reply torch_xla/csrc/elementwise.cpp:512]: Good point, we probably don't -- just having this call as a standard for safety.

**Review by bhavya01 — APPROVED** [review]

**Pull request merged**